### PR TITLE
Use Azure.Identity AzCliTokenProvider to get a token from AzCli

### PR DESCRIPTION
**Pull request by yogilad**
#### Pull Request Description

Scrap the proprietary AzCli implementation in favor of `Azure.Identity AzCliCredential`

#### Future Release Comment
**Breaking Changes:**
- Scrap the proprietary AzCli implementation in favor of `Azure.Identity AzCliCredential`.
Note this change requires AzCli version 2.3.0 and above installed to work.

### CODE REVIEW

File: azure-kusto-data/azure/kusto/data/_token_providers.py [hunk 1]
```diff
@@ -281,81 +277,42 @@ def _init_impl(self):
...
-    def _get_azure_cli_auth_token() -> dict:
-        """
-        Try to get the az cli authenticated token
-        :return: refresh token
-        """
-        import os
+            self._az_token = self._az_auth_context.get_token(self._kusto_uri)
+            return {TokenConstants.AZ_TOKEN_TYPE: TokenConstants.BEARER_TYPE, TokenConstants.AZ_ACCESS_TOKEN: self._az_token.token}
+        except Exception as e:
+            raise KustoClientError("Failed to obtain Az Cli token for '{0}'\n{1}".format(self._kusto_uri, e))
```
Comment (ArielYehezkely) [inline azure-kusto-data/azure/kusto/data/_token_providers.py:287]: Any suggestion how to solve this? The previous error message suggested to perform `az login`

File: azure-kusto-data/azure/kusto/data/_token_providers.py [hunk 2]
```diff
@@ -281,81 +277,42 @@ def _init_impl(self):
...
-            token_data = profile.get_raw_token()[0][2]
+            if self._az_auth_context_async is None:
+                self._az_auth_context_async = AsyncAzureCliCredential()
 
-            return token_data
+            self._az_token = await self._az_auth_context_async.get_token(self._kusto_uri, self._az_kwargs)
+            return {TokenConstants.AZ_TOKEN_TYPE: TokenConstants.BEARER_TYPE, TokenConstants.AZ_ACCESS_TOKEN: self._az_token.token}
+        except Exception as e:
+            raise KustoClientError(
+                "Failed to obtain Az Cli token for '{0}'.\nPlease be sure to use AzCli version 2.3.0 and above.\n{1}".format(self._kusto_uri, e)
```
Comment (ArielYehezkely) [inline azure-kusto-data/azure/kusto/data/_token_providers.py:298]: Is this true just for the async version?

File: azure-kusto-data/azure/kusto/data/_token_providers.py [hunk 3]
```diff
@@ -281,81 +277,42 @@ def _init_impl(self):
...
+                self._az_auth_context = AzureCliCredential()
 
-    @staticmethod
-    def _get_azure_cli_auth_token() -> dict:
-        """
-        Try to get the az cli authenticated token
-        :return: refresh token
-        """
-        import os
+            self._az_token = self._az_auth_context.get_token(self._kusto_uri)
```
Comment (ArielYehezkely) [inline azure-kusto-data/azure/kusto/data/_token_providers.py:284]: Do you know if it requires azure-cli package?

Comment (yogilad) [reply azure-kusto-data/azure/kusto/data/_token_providers.py:284]: This requires az cli to be installed, as it runs az-cli as a program.

File: azure-kusto-data/azure/kusto/data/_token_providers.py (hunk 2 above)
Comment (yogilad) [reply azure-kusto-data/azure/kusto/data/_token_providers.py:298]: What do you mean by _this_?

Comment (yogilad) [reply azure-kusto-data/azure/kusto/data/_token_providers.py:298]: No. Fixing...

**Review by ArielYehezkely — APPROVED** [review]

File: azure-kusto-data/azure/kusto/data/_token_providers.py (hunk 1 above)
Comment (yogilad) [reply azure-kusto-data/azure/kusto/data/_token_providers.py]: P.S. 
Regarding `az login`, AzCliCredential throws a specific exception message if not logged in.

File: azure-kusto-data/azure/kusto/data/_token_providers.py [hunk 4]
```diff
@@ -281,81 +277,44 @@ def _init_impl(self):
...
-                token_path = os.path.join(folder, "accessTokens.json")
-                with open(token_path) as f:
-                    data = json.load(f)
+        return None
 
-                # TODO: not sure I should take the first
-                return data[0]
-            except Exception as e:
-                raise KustoClientError("Azure cli token was not found. Please run 'az login' to setup account.", e)
+    async def _get_token_from_cache_impl_async(self) -> Optional[dict]:
```
Comment (AsafMah) [inline azure-kusto-data/azure/kusto/data/_token_providers.py:312]: This method is not needed - the default behavior of _get_token_from_cache_impl_async is to call the sync version

**Review by AsafMah — APPROVED** [review]

**Pull request merged**